### ISSUE TITLE
Provide documentation hosting and integrate with `cargo doc`

### ISSUE THREAD

**Issue by SimonSapin** [rust-lang/crates.io#91]
TL;DR: `cargo publish` should default to also running `cargo doc` and publishing the output online.

Having a crate’s documentation like http://doc.rust-lang.org/std/ available online is immensely valuable. Not only when using it, but also when deciding whether to use it (and so before `cargo doc` can be run locally).

Therefore, every crate on crates.io would ideally have its documentation online. For that to happen, it should be as easy as possible, with minimal effort from the developer.

The defaults should just work. `cargo publish` already involves building and testing crates, so documentation can be built at the same time. Even without any doc-comment, the rustdoc output can be useful. When the source package is uploaded to crates.io, the built documentation can be uploaded at the same time and be made available at https://example.net/[package-name]/ or https://example.net/[package-name]/[package-version]/, where example.net is a domain name managed together with crates.io. This URL should be linked from https://crates.io/crates/[package-name]

Although rustdoc is the default, if a project decides not to use rustdoc it should still be able to publish its documentation through this mechanism.

This effectively means providing free hosting of static files, which should be watched for abuse, but there is precedent: the Python Package Index does something similar with https://pythonhosted.org/. They have an HTTP API where one can upload a ZIP file, which the server exctracts to replace any previous content. Integration with Sphinx and distutils so that `python setup.py upload_docs` builds the documentation with Sphinx, zips it, and uploads it in one command.

Another case is Read The Docs, which pulls repositories and builds Sphinx documentation on its servers and can be trigerred on every commit by a GitHub hook. This is more risky as it involves running arbitrary code on the server.

If Cargo makes publishing/releasing a new version very easy, hopefully it can establish a convention in the Rust ecosystem of releasing often, so that updating the online docs for every commit (as opposed to releases) is not as necessary.

https://github.com/rust-lang/cargo/issues/1016 would work well together with this. If https://github.com/rust-lang/cargo/issues/1017 happens, it should be enabled for this.

Unresolved question: should the docs be made available for every version (at different URLs), or only the latest? If the former, should there be an unversioned URL for a default version? Is it a duplicate or a redirect? What navigation exists between versions? Read The Docs has a fixed-position overlay for this.

**Comment by andrewrk** [comment]
Is security the only concern preventing us from implementing this?

**Comment by nagisa** [comment]
Lack of an implementation is the biggest blocker. Getting a subdomain(s) or even domain ready for this is a very minor hurdle.

**Comment by Manishearth** [comment]
As far as security goes, all we need to do is write the workflow so that `cargo publish` can push it through `https://crates.io`, but the actual HTML is shown on `rustdoc.org` or something. `doc.crates.io` works too, one can deny cookie-sharing. The website hosting the docs is entirely static, no authentication whatsoever.

**Comment by nagisa** [comment]
@Manishearth one can easily deploy arbitrary JavaScript, and another can steal arbitrary cookies set by the first’s JS.

**Comment by Manishearth** [comment]
Yes, but that's not a security concern.

**Comment by SimonSapin** [comment]
IIRC cookie-related security concerns are why GitHub Pages and PyPI docs moved to github.io and pythonhosted.org. But that’s fine, we can use a separate domain as well.

**Comment by cmr** [comment]
rustdoc.org continues to be under my control and available.

**Comment by eternaleye** [comment]
There's also [docs.rs](http://docs.rs) - the owner has [explicitly stated](https://www.reddit.com/r/rust/comments/2wku7e/introducing_rustdoxcom_alpha_an_easy_way_to_host/corx5du) their willingness to pass it on to the core team.

**Comment by SimonSapin** [comment]
https://www.reddit.com/r/rust/comments/3fpo3r/youmu_online_rustdoc/ is an attempt at doing this with Travis-CI.

**Comment by Ticki** [comment]
:+1:

**Comment by filsmick** [comment]
:+1:

**Comment by teh** [comment]
+1 - I frequently run into out-of-date docs on self hosted static sites (e.g. github.io)

**Comment by rgardner** [comment]
Is there any work going on in this area? I like the approach godoc takes where you can visit `godoc.org/github.com/user/repo` and get the specific documentation for that user.

What prevents the current `travis-cargo doc-upload` from being modified to upload docs to `rustdoc.org` instead of GitHub pages? Is it a security concern due to arbitrary JavaScript like: `<script>alert(document.cookie)</script>`, or is it just that no one has implemented this yet?

**Comment by huonw** [comment]
@rgardner this would preferably be integrated with the versions (etc.) that are published to crates.io; godoc just shows the git HEAD (AFAIK) which may be arbitrarily different to the published versions, and travis-cargo certainly does.

**Comment by SimonSapin** [comment]
rustdoc.org doesn’t have any infrastructure yet.

**Comment by rgardner** [comment]
@huonw that makes sense that crates.io would have docs for each released version (instead of HEAD) and why using travis doesn't make sense for this problem.

So the blocker for this is the implementation hurdle of rolling our own GitHub Pages? (accepting static HTML/CSS/JS/font files?

**Comment by nagisa** [comment]
`rustdoc` output is not really suitable for proper documentation hosting service a-la [readthedocs.org](https://readthedocs.org/). `rustdoc`’s output only generates documentation for that particular package. In case of rustdoc.org/docs.rs/etc you’d probably want at least:

1. Cross-package search of some sort [a-la hoogle](http://hoogle.haskell.org/);
2. Easy way to switch between packages and their versions;
3. Proper crosslinking between packages;
3. etc etc etc.

None of these are, or can be, provided by `rustdoc` as it is currently.

**Comment by Manishearth** [comment]
> Cross-package search of some sort a-la hoogle;
>
> Proper crosslinking between packages;

Actually rustdoc does do cross package search with proper crosslinking, provided you run rustdoc with the doc folder containing the rustdoc output for the deps (which `cargo doc` does, iirc). See http://doc.servo.org/.

**Comment by nagisa** [comment]
> Actually rustdoc does do cross package search with proper crosslinking, provided you run rustdoc with the doc folder containing the rustdoc output for the deps (which cargo doc does, iirc). See http://doc.servo.org/. 

It does seem to work, yes. However, it is not a very good idea to send every client the search index of the whole crates.io ecosystem.

As for cross-linking, does it still work properly when package A depends on C 0.1 and B depends on C 1.1?

**Comment by Manishearth** [comment]
You don't need to send it to each client for the crosslinking to work. You sort of do, for Hoogle-like search, but perhaps that can be done by concatenating search.js (or whatever it's called)

**Comment by Ticki** [comment]
>  You sort of do, for Hoogle-like search

This could be an interesting project. Hoogle is definetly one of Haskell's strengths, I would love to see a similar thing for Rust. Rustdoc search still got much room for improvement.

**Comment by SimonSapin** [comment]
Rustdoc needs of lot of love, but that doesn't necessarily mean it should be dropped entirely.

And this is rather independent of what this issue is about: infrastructure to upload and host static files for a given crate and version.

**Comment by lfairy** [comment]
Is anyone currently working on this?

If not then I'd like to pick up the issue. I used to work on Hackage's doc hosting so I've got a good idea of the design space.

**Comment by Hywan** [comment]
:+1:

**Comment by alexcrichton** [comment]
@lfairy not currently, no, but we'd be excited to see movement here! We've talked about this in the past, and we likely don't want to commit to something too large just yet. Along those lines we've though:

* `cargo publish` will generate documentation for the host platform being published on
* The crate will be uploaded
* If successful, the documentation is also uploaded to a "default server"

We were thinking of using something like `rustdoc.org` for this where you'd just go to `rustdoc.org/$crate/$version` and you'd have all the rustdoc output. There's certainly lots more pieces to fill out in the long run, and even probably more pieces to fill out in the short term, but that's what we've thought will take us "most of the way there" with not a super high commitment.

**Comment by eternaleye** [comment]
@alexcrichton: One thing to consider, as documentation may vary with `#[cfg]`, is whether docs should be uploaded to `/$crate/$version/$hash_of_cfg_params`, and `/$crate/$version` redirects to one based on some logic, but shows a banner at the top showing dropdowns for cfg params to select a different one if the default is wrong (probably saving them in a cookie so other crates use the same parameters).

**Comment by alexcrichton** [comment]
@eternaleye yes a full blown "let's solve everyone's problems" solution to hosted docs would be significantly more complicated than what I outlined above. The idea is to that these kinds of features are the extra 10% which require a huge amount more work but don't bring *that* much benefit. 

We would arguably massively benefit from just providing *any* solution to hosting docs by default (in the sense that by default crates on crates.io have documentation hosted somewhere).

The idea is to also allow ourselves some enough leeway to implement any number of extensions in the future (such as various features, platform specific docs, etc).

**Comment by eternaleye** [comment]
@alexcrichton: My point is that the "first-take implementation" I laid out would be near-zero additional work on the server side (remember which one is newest for redirection), and very little overhead client-side (Cargo already has to know the cfg options for the conditional stuff that was added recently, so the only thing needed is formatting it and hashing, then sending the `(cfg_string, docs)` tuple to `/$crate/$version/$hash`). In exchange, it can be upgraded to a better system with 100% compatibility, rather than breaking URLs or having to hack around simply _not having_ the needed data. Because remember: Cool URIs don't change :stuck_out_tongue:

**Comment by alexcrichton** [comment]
Sure! So long as it's easy to implement seems reasonable to me

**Comment by onur** [comment]
Hi guys. I've been working on <https://crates.fyi> ([repo](https://github.com/onur/cratesfyi)) for a while and preparing a huge update. It's been fully functional since I announced it in reddit last month.

Current (outdated) version of crates.fyi is just pulling new releases from crates.io-index repository and generating documentations. I'll release an updated version of crates.fyi in few weeks, this update includes:

* Semver supported redirector to docs i.e: `rustdoc.crates.fyi/rand` will redirect to latest version, `rustdoc.crates.fyi/rand/^0.1` will redirect to latest point release of 0.1 etc.
* Simple clean UI to see long description (documentation of root of crate, and also TOC) and README.md if readme is defined in Cargo.toml
* Examples and source browser.
* Github stars and issues if a crate is hosted in github.
* Graphviz based deep dependency graph.
* Recent releases (similar to [metacpan](https://metacpan.org/recent)).
* List of crates by authors and owners.
* Similar crates (based on keyword).

And my future plans:

* Test results (cargo test output).
* Stable, beta, nightly build status (this requires constant building and
  a build queue).
* Code coverage (when it becomes more usable).
* Community maintained categories (like web frameworks in areweweb or something similar to awesome-rust).
* Code search.
* A documentation builder for github repositories similar to godoc.org.

It's mainly influenced by my favorite repository browser: metacpan and I've finished most of the stuff and started working on UI. I was reluctant about developing something like this (expected to see something official from rust team). But looks like you guys are really determined about pushing self hosted docs.

And ofc, build specific docs is a huge problem, but I believe we need organized hosted docs to give an overview to user about what he is looking. And if a user decides to use this crate, he can always build and use his own cfg/platform specific doc in future.

@lfairy if you still wondering if someone is working on this issue, I am working on it.

I'd like to hear your ideas. And whole project is written in rust, but since I am learning rust along with this project, its taking a bit time :-)

**Comment by Manishearth** [comment]
> Current (outdated) version of crates.fyi is just pulling new releases from crates.io-index repository and generating documentations

How do you handle sandboxing? Arbitrary code can get run during `cargo doc`.

I'd much prefer if `cargo publish` published docs built locally instead of having it done on the server. You don't have to worry about sandboxing and build times, then.


Godoc doesn't have this issue because go builds cannot run arbitrary code.

**Comment by onur** [comment]
I am building every crate in a chroot environment and using some pam_limits. It's pretty restricted environment and users in this build environment don't have any access to anything. I also prefer `cargo publish` if it's ever implemented. cratesfyi is just a tool to provide additional information (and currently missing hosted docs feature).

**Comment by nagisa** [comment]
Cross-platform documentation is a big pain point. You will almost never find documentation for e.g. Windows-specific modules. Whatever service spawns from this should support reading documentation from the perspective of at least all tier-1 platforms.

**Comment by alexcrichton** [comment]
@onur wow, what you've got planned sounds awesome! There's two possible things we may be able to do to help out as well:

1. Have docs uploaded on `cargo publish` directly to crates.fyi
2. Add webhooks on crates.io for when a crate is published.

That may be able to help out the multiple platform problem at least and also help keep docs up to date.

**Comment by steveklabnik** [comment]
Update: @onur has been doing a lot of work, and while it's not merged into mainline crates.fyi yet, it's shaping up really nicely.

**Comment by andrewdavidmackenzie** [comment]
+1 to having docs generated and published to crates.io as part of "cargo publish".

It would be very useful to be able to find a crate on crates.io and to browse it's documentation in the same site, before deciding to use it.

If there are other places it can be published, and either a command to do that ("cargo docs-publish"...), or "cargo docs" accepts an optional destination to publish to are all additional options.

**Comment by benaryorg** [comment]
@steveklabnik @onur is there any progress on this you can share?

**Comment by onur** [comment]
@benaryorg I forgot to update this thread, but check out recently released <https://docs.rs>.

**Comment by benaryorg** [comment]
Wow that looks beautiful.

If this fulfills all of the criteria then this issue could be closed I think.

**Comment by steveklabnik** [comment]
There's another issue discussing the integration, I would assume that if it
lands, it would close both this and that.

**Comment by carols10cents** [comment]
I'm going to close this since #676 has been merged and deployed!
